ocpmodels/common/relaxation/optimizers/lbfgs_torch.py

```diff
@@ -98,29 +101,32 @@ def run(self, fmax, steps):
                 ase.io.Trajectory(self.traj_dir / f"{name}.traj_tmp", mode="w")
                 for name in self.traj_names
             ]
+            # saves out the initial frame
+            self.write(e0, f0, update_mask, trajectories)
 
         iteration = 0
         converged = False
-        while iteration < steps and not converged:
+        while iteration < steps - 1 and not converged:
             r0, f0, e0 = self.step(
                 iteration, r0, f0, H0, rho, s, y, update_mask
             )
-            iteration += 1
-            if trajectories is not None:
-                self.atoms.y, self.atoms.force = e0, f0
-                atoms_objects = batch_to_atoms(self.atoms)
-                update_mask_ = torch.split(
-                    update_mask, self.atoms.natoms.tolist()
-                )
-                for atm, traj, mask in zip(
-                    atoms_objects, trajectories, update_mask_
-                ):
-                    if mask[0]:
-                        traj.write(atm)
-            update_mask = self.check_convergence(
+            _update_mask = self.check_convergence(
                 iteration, update_mask, f0, fmax
             )
-            converged = torch.all(torch.logical_not(update_mask))
+            converged = torch.all(torch.logical_not(_update_mask))
+            iteration += 1
+
+            if trajectories is not None:
+                # Save out ASE atoms objects only if:
+                # - save_full=True, saving out all frames else,
+                # - iteration==steps-1 or converged=True, saves out the converged/final frame, whichever comes first.
+                if self.save_full or converged or iteration == steps - 1:
+                    self.write(e0, f0, update_mask, trajectories)
```

Comment:
Aren't the positions that get written here after the update, while `e0` and `f0` are from before the update?